docs/adr/adr-014-bootstrap-from-previous-peers.md

```diff
@@ -0,0 +1,474 @@
+        }
+        return ex, nil
+}
+ ```
+
+(_Code Snippet 2.b: Example of required changes to: go-header/p2p/exchange.go_)
+
+The event handlers to be supplied are callbacks that either:
+
+1. Put the new peer list into the peer store
+2. Remove a blocked peer from the peer store
+
+Such callbacks are easily passable as options at header module construction time, example:
+
+```diff
++   fx.Provide(newPeerStore),
+    fx.Provide(newHeaderService)
+...
+- func(cfg Config, network modp2p.Network) []p2p.Option[p2p.ClientParameters] {
++ func(cfg Config, network modp2p.Network, peerStore datastore.Datastore) []p2p.Option[p2p.ClientParameters] {
+        return []p2p.Option[p2p.ClientParameters]{
+            ...
+            p2p.WithChainID(network.String()),
++           p2p.WithOnPeersGC(func(peers []peer.ID) {
++               var topTen []peer.ID
++               if len(peers) >= 10 {
++                  topTen = peers[:9]
++               } else {
++                   topTen = peers
++               }
++               peerStore.Put(topTen)
++           }),
+        }
+    },
+),
+                        fx.Provide(newP2PExchange(*cfg)),
+```
+
+(_Code Snippet 3.b: Example of required changes to celestia-node/nodebuilder/header/module.go_)
+
+As explained for _Code Snippet 1.b_ the event handler can handle how to deal with the ordered peer list, in this, we are portraying an example of situation where we are always interested in the top 10 peers if 10 or more peers are available.
+
+### Node Startup
+
+When the node starts up, it will first check if the `peers` database exists. If it does not, the node will bootstrap from the hardcoded bootstrappers. If it does, the node will bootstrap from the peers in the database.
+
+```diff
+ // newP2PExchange constructs a new Exchange for headers.
+ func newP2PExchange(cfg Config) func(
+        fx.Lifecycle,
+        ...
+        ...
+        []p2p.Option[p2p.ClientParameters],
++       pstore peerstore.Peerstore,
+ ) (libhead.Exchange[*header.ExtendedHeader], error) {
+        return func(
+                lc fx.Lifecycle,
+                ...
+                ...,
+                opts []p2p.Option[p2p.ClientParameters],
+        ) (libhead.Exchange[*header.ExtendedHeader], error) {
+                peers, err := cfg.trustedPeers(bpeers)
+                if err != nil {
+                        return nil, err
+                }
++              list, err := pstore.Load()
++               if err != nil {
++                       return nil, err
++               }
++               peers := append(peers, list...)
+                ids := make([]peer.ID, len(peers))
+                for index, peer := range peers {
+                        ids[index] = peer.ID
+                        host.Peerstore().AddAddrs(peer.ID, peer.Addrs, peerstore.PermanentAddrTTL)
+                }
+                exchange, err := p2p.NewExchange[*header.ExtendedHeader](host, ids, conngater, opts...)
```

Comment:
Here we take all the persistent peers and use them as trusted. However, trusted peers have a special trusted status for ___security___ reasons, and we can't take any random peers, even with high scores, and put them as trusted.

Trusted peers are exceptional because they are used for 'automatic subjective initialization'. Subjective initialization is part of any ProofOfStake protocol, and, in simple terms, it is bootstrapping of the current network validator set or header. Any user joining the network has to get the header somehow. Usually, in other protocols, users must manually get these headers/valset from trusted parties and manually provide them to their software. We went on a different, more user-friendly path where we request the initial headers from these trusted peers instead of forcing users to find them manually. Users are free to select custom trusted peers if they don't want to trust default bootstrappers and our infrastructure. 

If any random peer becomes trusted, it opens up a new vulnerability for go-header consumers. The malicious trusted peer can fool a user into following his fake fork by performing the Long Range Attack. This is why we cannot make persisted peers trusted and have to manage them separately.

Reply:
This concern has been addressed 👍🏻